File: src/leetcode/arrays_strings/26_remove_duplicates_from_sorted_array.ts
// LeetCode Top Interview 150
// https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/?envType=study-plan-v2&envId=top-interview-150

// Time complexity: O(n)
// Space complexity: O(1)
const removeDuplicates = (nums: number[]): number => {
  if (!nums) return 0;

  length = nums.length;

  if (length === 1) return 1;

  let uniqueNumCount = 1;

  for (let index = 1; index < length; index++) {
    const currentNum = nums[index];
    const lastUniqueNum = nums[uniqueNumCount - 1];

    if (lastUniqueNum === currentNum) continue;

    nums[uniqueNumCount] = currentNum;
    uniqueNumCount++;
  }

  return uniqueNumCount;
};

export default removeDuplicates;

File: src/leetcode/arrays_strings/26_remove_duplicates_from_sorted_array.rb
# frozen_string_literal: true

# LeetCode Top Interview 150
# https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/?envType=study-plan-v2&envId=top-interview-150

# Time complexity: O(n)
# Space complexity: O(1)

# @param {Integer[]} nums
# @return {Integer}
def remove_duplicates(nums)
  return 0 if nums.nil?

  length = nums.length

  return 1 if length == 1

  unique_num_count = 1

  nums.each do |current_num|
    last_unique_num = nums[unique_num_count - 1]

    next if last_unique_num == current_num

    nums[unique_num_count] = current_num
    unique_num_count += 1
  end

  unique_num_count
end
